### PR TITLE
[Plugin] check version of packages when install plugins

### PR DESCRIPTION
**Pull request by Humbertzhang**

### CODE REVIEW

File: skywalking/plugins/__init__.py [hunk 1]
```diff
@@ -38,7 +41,64 @@ def install():
...
+    '<': lambda cv, ev: cv < ev,
+    '<=': lambda cv, ev: cv < ev or cv == ev,
+    '=': lambda cv, ev: cv == ev,
+    '>=': lambda cv, ev: cv > ev or cv == ev,
+    '>': lambda cv, ev: cv > ev,
+    '!=': lambda cv, ev: cv != ev
+}
+
+
+def pkg_version_check(plugin, modname):
```
Comment (kezhenxu94) [inline skywalking/plugins/__init__.py:65]: Can you add some unit tests for this function? Just to check the `operators` work correctly

File: skywalking/plugins/__init__.py [hunk 2]
```diff
@@ -38,7 +41,64 @@ def install():
             continue
         logger.debug('installing plugin %s', modname)
         plugin = importer.find_module(modname).load_module(modname)
+
+        ok = pkg_version_check(plugin, modname)
+        if not ok:
```
[the remark this answers is not in the document]
Comment (Humbertzhang) [reply skywalking/plugins/__init__.py:46]: Is it ok to log `logger.debug('checking version for plugin %s', modname)` ?

File: skywalking/plugins/__init__.py [hunk 3]
```diff
@@ -38,7 +41,65 @@ def install():
             continue
         logger.debug('installing plugin %s', modname)
         plugin = importer.find_module(modname).load_module(modname)
+
+        logger.debug('checking version for plugin %s', modname)
+        supported = pkg_version_check(plugin, modname)
+        if not supported:
+            continue
```
Comment (kezhenxu94) [inline skywalking/plugins/__init__.py:48]: [1] move `logger.debug('checking version for plugin %s', modname)` into the `if not supported` block, and reword the message to tell the users that the current version is not supported by the plugin

File: skywalking/plugins/__init__.py [hunk 4]
```diff
@@ -38,7 +41,65 @@ def install():
...
+
+    pkg_name = plugin.version_rule.get("name")
+    rules = plugin.version_rule.get("rules")
+
+    try:
+        current_pkg_version = pkg_resources.get_distribution(pkg_name).version
+    except pkg_resources.DistributionNotFound:
+        supported = False
+        logger.warning("plugin %s didn\'t find the corresponding package %s, thus won't be installed",
+                       modname, pkg_name)
```
Comment (kezhenxu94) [inline skywalking/plugins/__init__.py:81]: When failed to get the version, we can just consider it as supported as the built-in libraries should always throw exception here

File: skywalking/plugins/__init__.py [hunk 5]
```diff
@@ -38,7 +41,65 @@ def install():
...
+        expect_pkg_version = rule[idx:]
+
+        current_version = version.parse(current_pkg_version)
+        expect_version = version.parse(expect_pkg_version)
+        f = _operators.get(symbol) or None
+
+        # version rule parse error, take it as no more version rules and return True.
+        if not f:
+            logger.warning("plugin %s version rule %s error. only allow >,>=,=,<=,<,!= symbols", modname, rule)
+            return supported
```
Comment (kezhenxu94) [inline skywalking/plugins/__init__.py:97]: We can simply raise an exception and exit immediately, this kind of exception should be fixed at development phase and never happened at runtime

File: skywalking/plugins/__init__.py [hunk 6]
```diff
@@ -38,7 +41,65 @@ def install():
...
+
+        if not f(current_version, expect_version):
+            supported = False
+            logger.warning("plugin %s need package %s version follow rules %s ,current version " +
+                           "is %s, thus won\'t be installed", modname, pkg_name, str(rules), current_pkg_version)
```
Comment (kezhenxu94) [inline skywalking/plugins/__init__.py:102]: Move this kind of message to where the invoker calls this method, I mark [1] in the previous comment

image

**Pull request merged**